### ISSUE TITLE
`Update pipeline` on a `paused pipeline` causes stuck merge

### ISSUE THREAD

**Issue by JoeyZwicker** [pachyderm/pachyderm#4294]
**What happened?**:
When a job is actively running with stats on:
1. stop pipeline
2. update pipeline
3. start pipeline

A new job will be triggered as expected but the previously running job will be stuck in running forever (cosmetically shown as running, it's actually finished behind the scenes I think -- shown below). The new job that get started gets stuck in merging.  I tested this on 1.9.8 and the "defensive closing of commits" code doesn't seem to fix this -- not sure if that was 1.9.8?

**What you expected to happen?**:
Current behavior for calling stop pipeline during a running job is fine. Pausing a pipeline shouldn't kill the job because user might start it back up. Perhaps there should be a "paused" state, but leaving in "running" is also reasonable. 

Just like calling `update pipeline` on a running job though, the existing job should be killed when update pipeline is called, regardless of the pipeline being stopped. Currently, we don't correctly clean up the `running` state. Stats commits also needs to be closed correctly. Right now it's not so next job gets stuck in merging. 

**How to reproduce it (as minimally and precisely as possible)?**:
Run opencv with `enable_stats:true`. Run a first job with the images.txt, then add images4.txt (has 40 more image added so it's a longer running job)

When a job is actively running with stats on:
1. stop pipeline edges
2. update pipeline edges
3. start pipeline edges

This will be the current state where things get stuck:

```
pc list job --history all
ID                               PIPELINE STARTED        DURATION           RESTART PROGRESS     DL       UL       STATE
9c0ae14108fa412ca4ae8fc2936246a6 edges    6 minutes ago  -                  0       24 + 18 / 42 3.314MiB 1.074MiB merging
897527bb65fa42c7aa52fd4fcc93329f montage  6 minutes ago  Less than a second 0       0 + 0 / 0    0B       0B       failure: inputs edges failed
56b382a63f6f40b8b8c500a99e5f3f60 edges    7 minutes ago  -                  0       12 + 0 / 42  6.466MiB 1.261MiB running
5f24e77fcac04bfe88a609562c404e13 montage  13 minutes ago 4 seconds          0       1 + 0 / 1    371.9KiB 1.283MiB success
dc6265e81fcc4ec2bd92063020c6adc5 montage  13 minutes ago 1 second           0       1 + 0 / 1    79.49KiB 378.6KiB success
28c1c713c6e64a2cb6a3730c5047f025 edges    13 minutes ago 3 seconds          0       2 + 1 / 3    181.1KiB 111.4KiB success
86eccdb950d946b49a091c322baa1db3 edges    13 minutes ago 2 seconds          0       1 + 0 / 1    57.27KiB 22.22KiB success
```

To confirm job stuck in running is just cosmetic:
```
pc stop job 56b382a63f6f40b8b8c500a99e5f3f60
error from StopJob: commit fb21a22aa24949b085b0d2cc5c9bec5e in repo edges has 
```
Stats commit left open
```
pc list commit edges
REPO  BRANCH COMMIT                           PARENT                           STARTED        DURATION       SIZE
edges stats  632f7b9315e9485591089fdcb33c40c4 32d536cfeab648b898cfe2555bf34a0d 20 minutes ago -              -
edges master 73031ea72377471c8f238924ef9d6e5e fb21a22aa24949b085b0d2cc5c9bec5e 20 minutes ago -              -
edges stats  32d536cfeab648b898cfe2555bf34a0d 3ab46d3b7c574d3d9149a27bb1b399b4 22 minutes ago -              -
edges master fb21a22aa24949b085b0d2cc5c9bec5e 4bd4457ff5184820816492d9e012f5f9 22 minutes ago About a minute 0B
edges stats  3ab46d3b7c574d3d9149a27bb1b399b4 c2a3af83d1434c138a2eb41399cd4bd6 28 minutes ago About a minute 380.4KiB
edges master 4bd4457ff5184820816492d9e012f5f9 cc8e968a8d284830a5194bd59bf46489 28 minutes ago About a minute 133.6KiB
edges stats  c2a3af83d1434c138a2eb41399cd4bd6 <none>                           28 minutes ago About a minute 83.49KiB
edges master cc8e968a8d284830a5194bd59bf46489 <none>                           28 minutes ago About a minute 22.22KiB
```

**Anything else we need to know?**:

**Environment?**:
- Kubernetes version (use `kubectl version`):
- Pachyderm CLI and pachd server version (use `pachctl version`): 1.9.8
- Cloud provider (e.g. aws, azure, gke) or local deployment (e.g. minikube vs dockerized k8s):
- OS (e.g. from /etc/os-release):
- Others:

**Comment by ysimonson** [comment]
I think Matt Goldberg on slack is hitting the same issue. In case it's helpful, here's a panic triggered from `run pipeline` with the cluster in this broken state:

```
020-02-03T23:33:20Z INFO pfs.API.FinishCommit {"request":{"trees":[{"hash":"d3d9ff0d117c2fbbba0382a106aa1bb4cd6104c4241064eb8cee9d97cf2f90c4c1c40a6a0449a136573eb83f42c4ed5dd28c4d2a432088bc5aa2a1775267e85e"}],"size_bytes":1177947154}} []
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1e719e9]
goroutine 247 [running]:
github.com/pachyderm/pachyderm/src/server/pkg/collection.runSTM.func1.1(0xc0000bdce0)
        /pachyderm/src/server/pkg/collection/transaction.go:100 +0xb3
panic(0x2183d40, 0x3c32390)
        /usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/pachyderm/pachyderm/src/server/pfs/server.(*driver).finishOutputCommit(0xc0003857c0, 0xc0010ad100, 0x0, 0xc00000e1a8, 0x1, 0x1, 0x0, 0x46360c12, 0x2471840, 0xc00033ed90)
        /pachyderm/src/server/pfs/server/driver.go:1483 +0x29
github.com/pachyderm/pachyderm/src/server/pfs/server.(*apiServer).FinishCommitInTransaction(0xc0000e0600, 0xc0010ad100, 0xc00033f5e0, 0xc00033f5e0, 0x0)
        /pachyderm/src/server/pfs/server/api_server.go:215 +0x8f
github.com/pachyderm/pachyderm/src/server/pkg/transactionenv.(*directTransaction).FinishCommit(0xc00000e1a0, 0xc00033ed90, 0x40ca01, 0xc00000e1a0)
        /pachyderm/src/server/pkg/transactionenv/env.go:220 +0x79
github.com/pachyderm/pachyderm/src/server/pfs/server.(*apiServer).FinishCommit.func3(0x29a5e80, 0xc00000e1a0, 0x234f020, 0x2301801)
        /pachyderm/src/server/pfs/server/api_server.go:226 +0x3a
github.com/pachyderm/pachyderm/src/server/pkg/transactionenv.(*TransactionEnv).WithTransaction.func1(0xc0010ad100, 0xc0010ad100, 0xc00033f0a0)
        /pachyderm/src/server/pkg/transactionenv/env.go:337 +0x6b
github.com/pachyderm/pachyderm/src/server/pkg/transactionenv.(*TransactionEnv).WithWriteContext.func1(0x29ab480, 0xc00033f0a0, 0x0, 0x0)
        /pachyderm/src/server/pkg/transactionenv/env.go:354 +0x160
github.com/pachyderm/pachyderm/src/server/pkg/collection.runSTM.func1(0xc0000bdce0, 0x29ab480, 0xc00033f0a0, 0xc00044d200, 0x40ca00)
        /pachyderm/src/server/pkg/collection/transaction.go:108 +0xad
created by github.com/pachyderm/pachyderm/src/server/pkg/collection.runSTM
        /pachyderm/src/server/pkg/collection/transaction.go:94 +0x86
```

**Comment by brycemcanally** [comment]
@JoeyZwicker The defensive code was not comprehensive in 1.9.8. This should be addressed by #4443 which went in with 1.9.10.

@ysimonson  I don't think the issue you are seeing is the same issue. The panic you ran into came from a job that produced output (trees is not empty), but seems to be missing the commit in the finish commit request. My guess would be that this is related to stats, and we are somehow getting into a situation where the stats commit in the job info is nil (not sure how that can happen though).

**Comment by adelelopez** [comment]
I tested this again and it appears it has been fixed after all the pipeline fixes we've pushed lately. The panic mentioned by Yusuf is a different issue, but has been fixed as well.
